File: grpc-service/src/pillow/pillow.controller.ts
import { Controller, Get, Param, Post, Put, Delete, Body } from '@nestjs/common';
import { PillowService } from './pillow.service';
import { GrpcMethod } from '@nestjs/microservices';
import { Datas } from 'src/protos/pillow/Datas';
import { DataID } from 'src/protos/pillow/DataID';
import { ParamsToFind } from 'src/protos/pillow/ParamsToFind';
import { DataDto } from 'src/protos/pillow/DataDto';
import { Metadata, ServerUnaryCall } from '@grpc/grpc-js';
import { ParamToFind } from 'src/protos/pillow/ParamToFind';
import { Data } from 'src/protos/pillow/Data';
import { Empty } from 'src/protos/pillow/Empty';
import { AvgHeartRate } from 'src/protos/pillow/AvgHeartRate';
import { AvgStressLevel } from 'src/protos/pillow/AvgStressLevel';

@Controller('pillow')
export class PillowController {

    constructor(private readonly pillowService:PillowService) {}

    @GrpcMethod('Pillow', 'GetDatas')
    async getDatas(metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Datas> {
      const data = await this.pillowService.findAll();
      return data;
    }
  
    @GrpcMethod('Pillow', 'GetData')
    async getData(dataID: DataID,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Data> {
      const data = await this.pillowService.findOne(dataID);
      return data;
    }
  
    @GrpcMethod('Pillow', 'GetPillowsByStressRate')
    async getPillowsByStressRate(params: ParamToFind,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Datas> {
      const data = await this.pillowService.findByStressRate(params);
      return data;
    }
    
    @GrpcMethod('Pillow', 'GetPillowsByHeartRate')
    async getPillowsByHeartRate(params: ParamsToFind,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Datas> {
      const data = await this.pillowService.findByHeartRate(params);
      return data;
    }

    @GrpcMethod('Pillow', 'GetPillowsBySnoringRange')
    async getPillowsBySnoringRange(params: ParamsToFind,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Datas> {
      const data = await this.pillowService.findBySnoringRange(params);
      return data;
    }

    @GrpcMethod('Pillow', 'GetPillowsByRespirationRate')
    async getPillowsByRespirationRate(params: ParamsToFind,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Datas> {
      const data = await this.pillowService.findByRespirationRate(params);
      return data;
    }

    @GrpcMethod('Pillow', 'AddData')
    async addData(data: DataDto,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Empty> {
      const dataToReturn = await this.pillowService.create(data);
      return {};
    }

    @GrpcMethod('Pillow', 'UpdateData')	
    async updateData(data: DataDto,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Data> {
      const dataToReturn = await this.pillowService.update(data);
      return dataToReturn;
    }

    @GrpcMethod('Pillow', 'DeleteData')
    async deleteData(dataID: DataID,metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<Empty> {
      const data = await this.pillowService.delete(dataID);
      return {};
    }

    @GrpcMethod('Pillow', 'GetAvgHeartRate')
    async getAvgHeartRate(metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<AvgHeartRate> {
      const data = await this.pillowService.getAvgHeartRate();
      return data;
    }
    @GrpcMethod('Pillow', 'GetAvgStressLevel')
    async getAvgStressLevel(metadata: Metadata, call: ServerUnaryCall<any, any>): Promise<AvgStressLevel> {
      const data = await this.pillowService.getAvgStressLevel();
      return data;
    }

   
    
}

File: ServiceNet/Controllers/PillowController.cs
using Microsoft.AspNetCore.Mvc;
using ServiceNet.Services;

namespace ServiceNet.Controllers
{
    [ApiController]
    [Route("[controller]")]
    public class PillowController : ControllerBase
    {
        private readonly IPillowService _pillowService;

        private readonly ILogger<PillowController> _logger;

        //public PillowController(ILogger<PillowController> logger)
        //{
        //    _logger = logger;
        //    _pillowService = PillowService.Instance;
        //}
        public PillowController(IPillowService pillowService)
        {
            _pillowService = pillowService;
        }

        [HttpPost("AddData")]
        public async Task<IActionResult> AddData([FromBody] DataDto request)
        {
            try
            {
                await _pillowService.AddData(request);
                return Ok();
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpDelete("DeleteData")]
        public async Task<IActionResult> DeleteData([FromBody] DataID request)
        {
            try
            {
                await _pillowService.DeleteData(request);
                return Ok();
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpGet("GetAvgHeartRate")]
        public async Task<IActionResult> GetAvgHeartRate()
        {
            try
            {
                var result = await _pillowService.GetAvgHeartRate(new Empty());
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpGet("GetAvgStressLevel")]
        public async Task<IActionResult> GetAvgStressLevel()
        {
            try
            {
                var result = await _pillowService.GetAvgStressLevel(new Empty());
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpGet("GetData")]
        public async Task<IActionResult> GetData([FromBody] DataID request)
        {
            try
            {
                var result = await _pillowService.GetData(request);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpGet("GetDatas")]
        public async Task<IActionResult> GetDatas()
        {
            try
            {
                var result = await _pillowService.GetDatas(new Empty());
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpGet("GetPillowsByHeartRate")]
        public async Task<IActionResult> GetPillowsByHeartRate([FromBody]ParamsToFind request)
        {
            try
            {
                var result = await _pillowService.GetPillowsByHeartRate(request);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpGet("GetPillowsByRespirationRate")]
        public async Task<IActionResult> GetPillowsByRespirationRate([FromBody]ParamsToFind reqest)
        {
            try
            {
                var result = await _pillowService.GetPillowsByRespirationRate(reqest);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpPost("GetPillowsBySnoringRange")]
        public async Task<IActionResult> GetPillowsBySnoringRange([FromBody] ParamsToFind request)
        {
            try
            {
                var result = await _pillowService.GetPillowsBySnoringRange(request);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpPost("GetPillowsByStressRate")]
        public async Task<IActionResult> GetPillowsByStressRate([FromBody] ParamToFind request)
        {
            try
            {
                var result = await _pillowService.GetPillowsByStressRate(request);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

        [HttpPost("UpdateData")]
        public async Task<IActionResult> UpdateData([FromBody] Data request)
        {
            try
            {
                var result = await _pillowService.UpdateData(request);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, ex.Message);
            }
        }

    }
}